rocAL/include/readers/image/numpy_data_reader.h

```diff
@@ -0,0 +1,149 @@
+    const std::string file_path() override { return _last_file_path; }
+
+    unsigned count_items() override;
+
+    ~NumpyDataReader() override;
+
+    int close() override;
+
+    NumpyDataReader();
+
+    size_t last_batch_padded_size() override;  // The size of the number of samples padded in the last batch
+
+    std::string get_root_folder_path() override;  // Returns the root folder path
+
+    std::vector<std::string> get_file_paths_from_meta_data_reader() override;  // Returns the relative file path from the meta-data reader
+   private:
+    //! opens the folder containnig the images
+    Reader::Status open_folder();
+    Reader::Status subfolder_reading();
+    std::string _folder_path;
+    std::string _file_list_path;
+    DIR *_src_dir;
+    DIR *_sub_dir;
+    struct dirent *_entity;
+    std::vector<std::string> _file_names;
+    std::vector<std::string> _files;
+    std::vector<NumpyHeaderData> _file_headers;
+    unsigned _curr_file_idx;
+    FILE *_current_fPtr;
+    unsigned _current_file_size;
+    unsigned _shard_start_idx;
+    std::vector<unsigned> _shard_start_idx_vector, _shard_end_idx_vector;
+    NumpyHeaderData _curr_file_header;
+    std::string _last_id;
+    std::string _last_file_name, _last_file_path, _absolute_file_path;
+    size_t _shard_id = 0;
+    size_t _shard_count = 1;  // equivalent of batch size
+    int32_t _shard_size = -1;
+    size_t _batch_size = 1;
+    size_t _padded_samples = 0;
+    bool _loop;
+    bool _shuffle;
+    int _read_counter = 0;
+    unsigned _seed = 0;
+    //!< _file_count_all_shards total_number of files in to figure out the max_batch_size (usually needed for distributed training).
+    size_t _file_count_all_shards;
+    std::mutex _cache_mutex;
+    std::map<std::string, NumpyHeaderData> _header_cache;
+    const RocalTensorDataType get_dtype(const std::string& format);
+    inline void skip_spaces(const char*& ptr);
+    void parse_header_data(NumpyHeaderData& target, const std::string& header);
+    template <size_t N>
+    void skip_char(const char*& ptr, const char (&what)[N]);
+    template <size_t N>
+    bool try_skip_char(const char*& ptr, const char (&what)[N]);
+    template <size_t N>
+    void skip_field(const char*& ptr, const char (&name)[N]);
+    template <typename T = int64_t>
+    T parse_int(const char*& ptr);
+    std::string parse_string(const char*& input, char delim_start = '\'', char delim_end = '\'');
+    void parse_header(NumpyHeaderData& parsed_header, std::string file_path);
+    template <typename T>
+    size_t parse_numpy_data(T* buf, std::vector<unsigned> strides, std::vector<unsigned> shapes, unsigned dim = 0);
+    bool get_header_from_cache(const std::string& file_name, NumpyHeaderData& target);
+    void update_header_cache(const std::string& file_name, const NumpyHeaderData& value);
+    void incremenet_read_ptr();
+    void increment_curr_file_idx();
+    int release();
+    void fill_last_batch();
```

Comment:
Is fille_lasth_batch and replicate_last_batch_to_pad_partial_shard used in .cpp files, please remove since latest LBP was accomodated.